src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -576,10 +578,12 @@ internal MemberAccessor MemberAccessorStrategy
         }
 
         [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
-        internal void RootBuiltInConvertersAndTypeInfoCreator()
+        internal void InitializeForReflectionSerializer()
         {
+            // For threading cases, the state that is set here can be overwritten.
             RootBuiltInConverters();
-            _typeInfoCreationFunc ??= CreateJsonTypeInfo;
+            _typeInfoCreationFunc = CreateJsonTypeInfo;
+            IsInitializedForReflectionSerializer = true;
```

Comment:
does this need to be a volatile write?

Reply:
No; replacement semantics is fine when assigning a reference. Any contention should only occur on the default options instance during initialization and may cause more than one func to be created, but that was true before as well since there was no lock.